### PR TITLE
[FIXED JENKINS-47339] Add build-implied REBUILD permission.

### PR DESCRIPTION
**Pull request by abayer**
[JENKINS-47339](https://issues.jenkins-ci.org/browse/JENKINS-47339)

Allows rebuilding a Pipeline run with the exact same script.

This will be used directly by Blue Ocean for its rebuild functionality, which doesn't need to be able to make changes to the script, so shouldn't require `Item.CONFIGURE` or `ReplayAction.REPLAY`. Instead, it'll require either `Item.BUILD` or `ReplayAction.REBUILD`.

cc @reviewbybees @i386

### CODE REVIEW

**Comment by jtnord** [comment]
> This will be used directly by Blue Ocean for its rebuild functionality, which doesn't need to be able to make changes to the script

Why does it just not trigger a build then?
Most of the times bo does this it is in fact wrong and should just ask for a build. 
https://issues.jenkins-ci.org/browse/JENKINS-42353

**Comment by abayer** [comment]
@jtnord To retain the parameters and SCM revision.

**Comment by abayer** [comment]
Seems like repo.azure.jenkins.io has a busted cache of the 2.73.1 war file. Fun!

**Comment by abayer** [comment]
@jglick Fair enough.

File: src/test/java/org/jenkinsci/plugins/workflow/cps/replay/ReplayActionTest.java [hunk 1]
```diff
@@ -189,6 +195,15 @@ private static boolean canReplay(WorkflowRun b, String user) {
         });
     }
 
+    private static boolean canRebuild(WorkflowRun b, String user) {
+        final ReplayAction a = b.getAction(ReplayAction.class);
+        return ACL.impersonate(User.get(user).impersonate(), new NotReallyRoleSensitiveCallable<Boolean,RuntimeException>() {
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/replay/ReplayActionTest.java]: BTW `ACL.as` is generally easier to use.

File: src/test/java/org/jenkinsci/plugins/workflow/cps/replay/ReplayActionTest.java [hunk 2]
```diff
@@ -271,4 +286,51 @@ private static boolean canReplay(WorkflowRun b, String user) {
...
+                    story.j.waitUntilNoActivity();
+                    b2 = p.getBuildByNumber(2);
+                    assertNotNull(b2);
+                }
+                story.j.assertLogContains("script to rebuild", story.j.assertBuildStatusSuccess(b2));
+                ReplayCause cause = b2.getCause(ReplayCause.class);
+                assertNotNull(cause);
+                assertEquals(1, cause.getOriginalNumber());
+                assertEquals(b1, cause.getOriginal());
+                assertEquals(b2, cause.getRun());
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/workflow/cps/replay/ReplayActionTest.java]: Might be interesting to verify that, say, parameters are the same as in the original.

File: src/main/resources/org/jenkinsci/plugins/workflow/cps/replay/ReplayAction/index.jelly [hunk 3]
```diff
@@ -4,27 +4,38 @@
     <l:layout title="Replay ${it.owner.displayName}" norefresh="true">
         <st:include page="sidepanel.jelly" it="${it.owner}"/>
         <l:main-panel>
-            <j:if test="${it.enabled}">
+            <j:if test="${it.enabled or it.rebuildEnabled}">
                 <h1>Replay ${it.owner.displayName}</h1>
                 <p>
                     <j:out value="${%blurb}"/>
                 </p>
-                <f:form action="run" method="POST" name="config">
```
Comment (jglick) [inline src/main/resources/org/jenkinsci/plugins/workflow/cps/replay/ReplayAction/index.jelly]: [diff for reviewers](https://github.com/jenkinsci/workflow-cps-plugin/pull/187/files?w=1#diff-bea9feb20818b01fa16e0538ec82cccfR12)

**Pull request merged**